### PR TITLE
check: add precheck lib

### PR DESCRIPTION
**Pull request by IANTHEREAL**
* refine some old pkg/db codes
* add db precheck functions

### CODE REVIEW

**Comment by IANTHEREAL** [comment]
@csuzhangxc PTAL

**Comment by csuzhangxc** [comment]
@GregoryIan fix CI first?

**Comment by IANTHEREAL** [comment]
I would add test later for them

File: pkg/dbutil/query.go [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+)
+
+// QuerySQL queries sql, and returns some row
+func QuerySQL(db *sql.DB, query string) (*sql.Rows, error) {
+	log.Debugf("[query][sql] %s", query)
+	rows, err := db.Query(query)
+	if err != nil {
+		log.Errorf("query sql[%s] failed %v", query, errors.ErrorStack(err))
+		return nil, errors.Trace(err)
+	}
```
Comment (csuzhangxc) [inline pkg/dbutil/query.go]: maybe we can just return the error but do not log it and let outer caller to log the error?

Comment (IANTHEREAL) [reply pkg/dbutil/query.go]: yep, I like it, let's define a log specification

File: pkg/precheck/precheck.go [hunk 2]
```diff
@@ -0,0 +1,123 @@
+package precheck
+
+import (
+	"sync"
+
+	"github.com/juju/errors"
+	"github.com/ngaut/log"
+)
+
+// PreChecker is interface about precheck.
```
Comment (csuzhangxc) [inline pkg/precheck/precheck.go]: `// PreChecker is the interface about precheck.`
and some comments below ditto

File: pkg/precheck/precheck.go [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+	PreCheck() *PreCheckResult
+}
+
+// PreCheckState is state of precheck
+type PreCheckState string
+
+const (
+	PreCheckState_Success PreCheckState = "success"
+	PreCheckState_Failure PreCheckState = "fail"
+	PreCheckState_Warning PreCheckState = "warn"
```
Comment (csuzhangxc) [inline pkg/precheck/precheck.go]: how about
```
StateSuccess PreCheckState = "success"
StateFailure PreCheckState = "fail"
StateWarning PreCheckState = "warn"
```
i think we don't need the variable name starts with the package name

and below ditto, `PreCheckResults`, `PreCheckResultSummary`, `PreCheckResults`

File: pkg/precheck/precheck.go [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+	Extra       string        `json:"extra"`
+}
+
+// PreCheckResultSummary is summary of all precheck results
+type PreCheckResultSummary struct {
+	Passed     bool  `json:"passed"`
+	Total      int64 `json:"total"`
+	Successful int64 `json:"successful"`
+	Failed     int64 `json:"failed"`
+	Warning    int64 `son:"warning"`
```
Comment (csuzhangxc) [inline pkg/precheck/precheck.go]: Tyno, `json:"warning"`

File: pkg/precheck/precheck.go [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+					failed++
+				case PreCheckState_Warning:
+					warning++
+				}
+
+				// if total == successful + warning + failed, it's finished
+				finished = (total == successful+warning+failed)
+				results.Results = append(results.Results, result)
+
+				log.Debugf("[precheck] precheck request %s result:%+v", result, result)
```
Comment (csuzhangxc) [inline pkg/precheck/precheck.go]: log `result` twice?

File: pkg/precheck/precheck.go [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+			}
+		}
+	}()
+
+	for i, prechecker := range precheckers {
+		wg.Add(1)
+		go func(i int, prechecker PreChecker) {
+			defer wg.Done()
+			result := prechecker.PreCheck()
+			result.Id = uint64(i)
```
Comment (csuzhangxc) [inline pkg/precheck/precheck.go]: seems `Id` is not useful

File: pkg/precheck/precheck.go [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+}
+
+// PreCheckResults contains all precheck results and summary
+type PreCheckResults struct {
+	Results []*PreCheckResult      `json:"results"`
+	Summary *PreCheckResultSummary `json:"summary"`
+}
+
+// DoPreCheck executes several precheckers.
+func DoPreCheck(precheckers []PreChecker) (*PreCheckResults, error) {
```
Comment (csuzhangxc) [inline pkg/precheck/precheck.go]: do we need to add `ctx` to suppoer cancel

File: pkg/check/precheck.go [hunk 8]
```diff
@@ -0,0 +1,126 @@
+package check
+
+import (
+	"sync"
+
+	"github.com/juju/errors"
+	"github.com/ngaut/log"
+)
+
+// Checker is interface that defines checker.
```
Comment (csuzhangxc) [inline pkg/check/precheck.go]: `// Checker is the interface that defines checker.`
and some comments below ditto

File: pkg/check/precheck.go [hunk 9]
```diff
@@ -0,0 +1,126 @@
...
+	Extra       string `json:"extra"`
+}
+
+// ResultSummary is summary of all check results
+type ResultSummary struct {
+	Passed     bool  `json:"passed"`
+	Total      int64 `json:"total"`
+	Successful int64 `json:"successful"`
+	Failed     int64 `json:"failed"`
+	Warning    int64 `son:"warning"`
```
Comment (csuzhangxc) [inline pkg/check/precheck.go]: Tyno, `json:"warning"`

File: pkg/check/precheck.go [hunk 10]
```diff
@@ -0,0 +1,126 @@
...
+			}
+		}
+	}()
+
+	for i, checker := range checkers {
+		wg.Add(1)
+		go func(i int, checker Checker) {
+			defer wg.Done()
+			result := checker.Check()
+			result.ID = uint64(i)
```
Comment (csuzhangxc) [inline pkg/check/precheck.go]: seems `ID` is not useful

File: pkg/check/precheck.go [hunk 11]
```diff
@@ -0,0 +1,126 @@
...
+}
+
+// Results contains all check results and summary
+type Results struct {
+	Results []*Result      `json:"results"`
+	Summary *ResultSummary `json:"summary"`
+}
+
+// Do executes several checkers.
+func Do(checkers []Checker) (*Results, error) {
```
Comment (csuzhangxc) [inline pkg/check/precheck.go]: do we need to add `ctx` to suppoer cancel

Comment (IANTHEREAL) [reply pkg/check/precheck.go]: good catch

File: pkg/check/precheck.go (hunk 10 above)
Comment (IANTHEREAL) [reply pkg/check/precheck.go]: I'm not sure about this, but I'll keep it in case someone uses it

File: pkg/check/binlog.go [hunk 12]
```diff
@@ -0,0 +1,161 @@
...
+	"fmt"
+	"strings"
+
+	"github.com/juju/errors"
+	"github.com/ngaut/log"
+	"github.com/pingcap/tidb-tools/pkg/dbutil"
+)
+
+// MySQLBinlogEnablePreChecker checks whether `log_bin` variable is enabled in MySQL.
+type MySQLBinlogEnablePreChecker struct {
```
Comment (csuzhangxc) [inline pkg/check/binlog.go]: rename `MySQLBinlogEnablePreChecker` to `MySQLBinlogEnableChecker`, and the same to `MySQLBinlogFormatPreChecker`, `MySQLBinlogRowImagePreChecker`, etc.

File: pkg/check/binlog.go [hunk 13]
```diff
@@ -141,7 +142,7 @@ func (pc *MySQLBinlogRowImagePreChecker) Check() *Result {
 		return result
 	}
 
-	value, err = dbutil.ShowBinlogRowImage(pc.db)
+	value, err = dbutil.ShowBinlogRowImage(ctx, pc.db)
 	if err != nil {
 		result.ErrorMsg = errors.ErrorStack(err)
```
Comment (csuzhangxc) [inline pkg/check/binlog.go]: use `markCheckError` instead?

File: sync_diff_inspector/chunk.go [hunk 14]
```diff
@@ -67,7 +68,7 @@ func getChunksForTable(db *sql.DB, Schema, tableName string, column *model.Colum
 		"/*!40001 SQL_NO_CACHE */", field, field, Schema, tableName, where)
 
 	// get the chunk count
-	cnt, err := pkgdb.GetRowCount(db, Schema, tableName, where)
+	cnt, err := dbutil.GetRowCount(context.Background(), db, Schema, tableName, where)
```
Comment (IANTHEREAL) [inline sync_diff_inspector/chunk.go:71]: @WangXiangUSTC please refine your process flow later

File: pkg/check/check.go [hunk 15]
```diff
@@ -0,0 +1,128 @@
...
+	total = int64(len(checkers))
+
+	resultCh := make(chan *Result)
+
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		for {
+			select {
+			case result := <-resultCh:
```
Comment (csuzhangxc) [inline pkg/check/check.go:82]: how about `for result := range resultCh` ?

Comment (csuzhangxc) [inline pkg/check/check.go]: how about `for result := range resultCh` ?

File: pkg/check/check.go [hunk 16]
```diff
@@ -0,0 +1,128 @@
...
+	results.Summary = &ResultSummary{
+		Passed:     passed,
+		Total:      total,
+		Successful: successful,
+		Failed:     failed,
+		Warning:    warning,
+	}
+
+	log.Infof("check finished, passed %v / total %v", passed, total)
+	return results, errors.Trace(err)
```
Comment (csuzhangxc) [inline pkg/check/check.go]: `err` seems always is `nil`, keep it for expansibility？

File: pkg/check/mysql_server.go [hunk 17]
```diff
@@ -0,0 +1,107 @@
...
+
+// Name implements the Checker interface.
+func (pc *MySQLVersionChecker) Name() string {
+	return "mysql_version"
+}
+
+/*****************************************************/
+
+// MySQLServerIDChecker checks mysql/mariadb version.
+type MySQLServerIDChecker struct {
```
Comment (csuzhangxc) [inline pkg/check/mysql_server.go]: checks mysql/mariadb server_id

File: pkg/check/privilege.go [hunk 18]
```diff
@@ -0,0 +1,121 @@
...
+
+	var (
+		hasPrivilegeOfReplicationSlave  bool
+		hasPrivilegeOfReplicationClient bool
+		hasPrivilegeOfReload            bool
+	)
+	if len(grantStmt.Users) == 0 {
+		result.ErrorMsg = fmt.Sprintf("grant has not user %s", grantStmt.Text())
+		return result
+	}
```
Comment (csuzhangxc) [inline pkg/check/privilege.go]: little suggestion, move block to ahead of `var (`

File: pkg/check/util.go [hunk 19]
```diff
@@ -0,0 +1,76 @@
...
+// mysql        5.7.18-log
+// mariadb      5.5.50-MariaDB-1~wheezy
+// percona      5.7.19-17-log
+// aliyun rds   5.7.18-log
+// aws rds      5.7.16-log
+// ref: https://dev.mysql.com/doc/refman/5.7/en/which-version.html
+
+// v is mysql version in string format.
+func toMySQLVersion(v string) MySQLVersion {
+	tmp := strings.Split(v, "-")
```
Comment (csuzhangxc) [inline pkg/check/util.go:27]: should we test `len(tmp) > 0` ?

File: pkg/check/util.go [hunk 20]
```diff
@@ -0,0 +1,76 @@
...
+	tmp = strings.Split(tmp[0], ".")
+	if len(tmp) != 3 {
+		log.Warnf("invalid version %s", v)
+		return [3]uint{0, 0, 0}
+	}
+	version := [3]uint{}
+	for i := range tmp {
+		v, err := strconv.ParseUint(tmp[i], 10, 64)
+		if err != nil {
+			log.Warnf("invalid version %s", v)
```
Comment (csuzhangxc) [inline pkg/check/util.go]: `log.Warnf("invalid version %s", tmp[i])` ?

File: pkg/dbutil/README.md [hunk 21]
```diff
@@ -0,0 +1,3 @@
+# dbutil library
+
+dbutil is a library that contains a collection of auxiliary function related to MySQL and TiDB.
```
Comment (csuzhangxc) [inline pkg/dbutil/README.md]: functions  ?

File: pkg/check/util.go (hunk 20 above)
Comment (IANTHEREAL) [reply pkg/check/util.go]: output complete string

File: pkg/check/check.go (hunk 15 above)
Comment (IANTHEREAL) [reply pkg/check/check.go:82]: both are ok

File: pkg/check/util.go (hunk 20 above)
Comment (csuzhangxc) [reply pkg/check/util.go]: `v` become new variable returned by `strconv.ParseUint` (an uint64)？
maybe we need a `v2` or another

File: pkg/check/check.go (hunk 15 above)
Comment (csuzhangxc) [reply pkg/check/check.go:82]: I know, maybe `for result := range resultCh` seems simpler, maybe

File: pkg/check/util.go (hunk 20 above)
Comment (IANTHEREAL) [reply pkg/check/util.go]: we need to output complete string :smiley_cat:

File: pkg/check/check.go (hunk 15 above)
Comment (IANTHEREAL) [reply pkg/check/check.go:82]: I mean If something can be changed or not, leave it to your favor :smile:

File: pkg/dbutil/common.go [hunk 22]
```diff
@@ -0,0 +1,266 @@
...
+func (c *DBConfig) String() string {
+	if c == nil {
+		return "<nil>"
+	}
+	return fmt.Sprintf("DBConfig(%+v)", *c)
+}
+
+// CreateDB creates a mysql connection FD
+func CreateDB(cfg DBConfig) (*sql.DB, error) {
+	dbDSN := fmt.Sprintf("%s:%s@tcp(%s:%d)/%s?charset=utf8", cfg.User, cfg.Password, cfg.Host, cfg.Port, cfg.Schema)
```
Comment (csuzhangxc) [inline pkg/dbutil/common.go]: should we need to support charset with `utf8mb4`

File: pkg/dbutil/common.go [hunk 23]
```diff
@@ -0,0 +1,266 @@
...
+	}
+	if !tbl.Valid || !createTable.Valid {
+		return "", errors.NotFoundf("table %s", tableName)
+	}
+	return createTable.String, nil
+}
+
+// GetRowCount returns row count of the table.
+// if not specify where condition, return total row count of the table.
+func GetRowCount(ctx context.Context, db *sql.DB, dbName string, table string, where string) (int64, error) {
```
Comment (csuzhangxc) [inline pkg/dbutil/common.go]: unify `dbName` to `schemaName`? ditto below

File: pkg/dbutil/common.go [hunk 24]
```diff
@@ -0,0 +1,266 @@
...
+		|  100 |
+		+------+
+	*/
+
+	query := fmt.Sprintf("SELECT COUNT(1) cnt FROM `%s`.`%s`", dbName, table)
+	if len(where) > 0 {
+		query += fmt.Sprintf(" WHERE %s", where)
+	}
+
+	rows, err := QuerySQL(ctx, db, query)
```
Comment (csuzhangxc) [inline pkg/dbutil/common.go]: how about use `db.QueryRowContext`, the same for below

File: pkg/dbutil/query.go [hunk 25]
```diff
@@ -0,0 +1,66 @@
...
+
+import (
+	"context"
+	"database/sql"
+
+	"github.com/juju/errors"
+	"github.com/ngaut/log"
+)
+
+// QuerySQL queries sql, and returns some row
```
Comment (csuzhangxc) [inline pkg/dbutil/query.go]: some rows

File: pkg/check/check.go [hunk 26]
```diff
@@ -0,0 +1,126 @@
...
+	Check(ctx context.Context) *Result
+}
+
+// State is state of check
+type State string
+
+const (
+	// StateSuccess indicates that the check was successful
+	StateSuccess State = "success"
+	// StateFailure indicates that the check was faiful
```
Comment (WangXiangUSTC) [inline pkg/check/check.go]: faiful? typo

File: pkg/check/privilege.go [hunk 27]
```diff
@@ -0,0 +1,122 @@
...
+// RELOAD is strongly suggested to have.
+func (pc *SourcePrivilegeChecker) Check(ctx context.Context) *Result {
+	result := &Result{
+		Name:  pc.Name(),
+		Desc:  "checks data source privileges",
+		State: StateFailure,
+		Extra: fmt.Sprintf("%s:%d", pc.dbinfo.Host, pc.dbinfo.Port),
+	}
+
+	grants, err := dbutil.ShowGrants(ctx, pc.db, "", "")
```
Comment (WangXiangUSTC) [inline pkg/check/privilege.go:41]: should we specify the user?

File: sync_diff_inspector/chunk.go (hunk 14 above)
Comment (WangXiangUSTC) [reply sync_diff_inspector/chunk.go:71]: ok

**Review by WangXiangUSTC — COMMENTED** [review]
rest LGTM

File: pkg/check/privilege.go (hunk 27 above)
Comment (IANTHEREAL) [reply pkg/check/privilege.go:41]: no need, we check current user/host (that are specified while we create db connection)

**Review by csuzhangxc — APPROVED** [review]
LGTM

**Pull request merged**